src/object/srv_obj.c

```diff
@@ -720,85 +1039,216 @@ ds_obj_rw_handler(crt_rpc_t *rpc)
 	struct ds_cont_hdl		*cont_hdl = NULL;
 	struct ds_cont			*cont = NULL;
+	struct daos_tx_handle		*dth = NULL;
+	struct daos_tx_id		*dti = NULL;
 	daos_handle_t			 ioh = DAOS_HDL_INVAL;
 	uint32_t			 map_ver = 0;
 	int				 tag;
 	bool				 update;
 	bool				 dispatch;
+	bool				 resend = false;
+	int				 count = 0;
 	int				 dispatch_rc = 0;
 	int				 rc;
+	int				 rc1;
 
 	D_ASSERT(orw != NULL);
 	D_ASSERT(orwo != NULL);
 	update = (opc_get(rpc->cr_opc) == DAOS_OBJ_RPC_UPDATE);
 	dispatch = update && orw->orw_shard_tgts.ca_arrays != NULL;
-	tag = dss_get_module_info()->dmi_tgt_id;
 
-	D_DEBUG(DB_TRACE, "rpc %p opc %d "DF_UOID" dkey %d %s tag/xs %d/%d eph "
-		DF_U64".\n", rpc, opc_get(rpc->cr_opc), DP_UOID(orw->orw_oid),
-		(int)orw->orw_dkey.iov_len, (char *)orw->orw_dkey.iov_buf,
-		tag, dss_get_module_info()->dmi_xs_id, orw->orw_epoch);
+	tag = dss_get_module_info()->dmi_tgt_id;
 	rc = ds_check_container(orw->orw_co_hdl, orw->orw_co_uuid,
 				&cont_hdl, &cont);
-	if (rc)
-		goto out;
+	if (rc != 0)
+		goto reply;
 
-	if (opc_get(rpc->cr_opc) == DAOS_OBJ_RPC_UPDATE &&
-	    !(cont_hdl->sch_capas & DAOS_COO_RW)) {
+	if (update && !(cont_hdl->sch_capas & DAOS_COO_RW)) {
 		D_ERROR("cont "DF_UUID" sch_capas "DF_U64", "
 			"NO_PERM to update.\n",
 			DP_UUID(orw->orw_co_uuid), cont_hdl->sch_capas);
-		D_GOTO(out, rc = -DER_NO_PERM);
+		D_GOTO(reply, rc = -DER_NO_PERM);
 	}
 
 	D_ASSERT(cont_hdl->sch_pool != NULL);
 	map_ver = cont_hdl->sch_pool->spc_map_version;
+
+	D_DEBUG(DB_TRACE, "rpc %p opc %d "DF_UOID" dkey %d %s tag/xs %d/%d eph "
+		DF_U64", pool ver %u/%u with "DF_DTI".\n",
+		rpc, opc_get(rpc->cr_opc), DP_UOID(orw->orw_oid),
+		(int)orw->orw_dkey.iov_len, (char *)orw->orw_dkey.iov_buf,
+		tag, dss_get_module_info()->dmi_xs_id, orw->orw_epoch,
+		orw->orw_map_ver, map_ver, DP_DTI(&orw->orw_dti));
+
 	if (orw->orw_map_ver < map_ver) {
 		D_DEBUG(DB_IO, "stale version req %d map_version %d\n",
 			orw->orw_map_ver, map_ver);
 		if (update && dispatch)
-			D_GOTO(out, rc = -DER_STALE);
+			D_GOTO(reply, rc = -DER_STALE);
+	} else if (orw->orw_map_ver > map_ver) {
+		/* XXX: client has newer pool map than the server.
+		 *	Refresh the pool map on the server?
+		 */
 	}
 
-	/* dispatch to other tgts when needed */
+	if (!update)
+		goto local;
+
+	if (daos_oc_echo_type(daos_obj_id2class(orw->orw_oid.id_pub)))
+		goto dispatch;
+
+	/* Check DTX configuration. */
+	if (orw->orw_flags & ORF_FROM_LEADER) {
+		if (((orw->orw_flags & ORF_DTX_DISABLED) == 0) !=
+		    srv_enable_dtx) {
+			D_ERROR("Inconsistent DTX configuration among "
+				"servers.\n");
+			D_GOTO(reply, rc = -DER_PROTO);
+		}
+	} else {
+		if (daos_is_zero_dti(&orw->orw_dti) && srv_enable_dtx) {
+			D_ERROR("Inconsistent DTX configuration between "
+				"client and server.\n");
+			D_GOTO(reply, rc = -DER_PROTO);
+		}
+	}
+
+	/* Handle resend. */
+	if (orw->orw_flags & ORF_RESEND) {
+		rc = obj_handle_resend(cont->sc_hdl, &orw->orw_oid,
```

Comment:
need not check if the DTX enabled? for example for client replica mode that disabled DTX

Reply:
You are right. It has been handled in the new push.